yarn-project/ethereum/src/deploy_l1_contracts.ts

```diff
@@ -651,15 +652,63 @@ export async function deployL1Contract(
     const existing = await publicClient.getBytecode({ address });
 
     if (existing === undefined || existing === '0x') {
-      txHash = await walletClient.sendTransaction({ to: deployer, data: concatHex([salt, calldata]) });
-      logger?.verbose(`Deploying contract with salt ${salt} to address ${address} in tx ${txHash}`);
+      // Add gas estimation and price buffering for CREATE2 deployment
+      const deployData = encodeDeployData({ abi, bytecode, args });
+      const gasUtils = new GasUtils(publicClient, logger);
+      const gasEstimate = await gasUtils.estimateGas(() =>
+        publicClient.estimateGas({
+          account: walletClient.account?.address,
+          data: deployData,
+        }),
+      );
+      const gasPrice = await gasUtils.getGasPrice();
+
+      txHash = await walletClient.sendTransaction({
+        to: deployer,
+        data: concatHex([salt, calldata]),
+        gas: gasEstimate,
+        maxFeePerGas: gasPrice,
+      });
+      logger?.verbose(
+        `Deploying contract with salt ${salt} to address ${address} in tx ${txHash} (gas: ${gasEstimate}, price: ${gasPrice})`,
+      );
     } else {
       logger?.verbose(`Skipping existing deployment of contract with salt ${salt} to address ${address}`);
     }
   } else {
-    txHash = await walletClient.deployContract({ abi, bytecode, args });
-    logger?.verbose(`Deploying contract in tx ${txHash}`);
-    const receipt = await publicClient.waitForTransactionReceipt({ hash: txHash, pollingInterval: 100 });
+    // Regular deployment path
+    const deployData = encodeDeployData({ abi, bytecode, args });
+    const gasUtils = new GasUtils(publicClient, logger);
+    const gasEstimate = await gasUtils.estimateGas(() =>
+      publicClient.estimateGas({
+        account: walletClient.account?.address,
+        data: deployData,
+      }),
+    );
+    const gasPrice = await gasUtils.getGasPrice();
+
+    txHash = await walletClient.deployContract({
+      abi,
+      bytecode,
+      args,
+      gas: gasEstimate,
+      maxFeePerGas: gasPrice,
+    });
+
+    // Monitor deployment transaction
+    await gasUtils.monitorTransaction(txHash, walletClient, {
+      to: '0x', // Contract creation
+      data: deployData,
+      nonce: await publicClient.getTransactionCount({ address: walletClient.account!.address }),
+      gasLimit: gasEstimate,
+      maxFeePerGas: gasPrice,
+    });
+
+    const receipt = await publicClient.waitForTransactionReceipt({
+      hash: txHash,
+      pollingInterval: 100,
+      timeout: 60_000, // 1min
+    });
```

Comment:
This won't work if the `gasUtils` has issued a replacement, since changing the gas price changes the tx hash, so you'll be listening for a different one. Easiest solution here is to poll on the `getTxCount` so you know via the nonce when any of all the replacement txs has been mined, and then do a `getReceipt` on every one of the tx hashes you've sent until you get a hit (at least until [`eth_getTransactionBySenderAndNonce`](https://github.com/ethereum/execution-apis/issues/494) is accepted and implemented).

Most likely you'll want to hide this within `GasUtils`.